Negative flow fencing

n1r1 opened the pull request:
Creating a dummy PR for easier review before PR against upstream.
This can be closed once we finish review

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -54,6 +54,16 @@ const (
+	machineRoleLabel                = "machine.openshift.io/cluster-api-machine-role"
+	machineRoleMaster               = "master"
+	// TODO: set meaningful defaults
+	annotationTimestampFormat          = time.RFC3339
```
There's a formatting issue here, `go fmt` test might not like it

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -883,6 +909,112 @@ func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMe
+}
+
+func (a *Actuator) tryDeleteMachine(ctx context.Context, machine *machinev1beta1.Machine, baremetalhost *bmh.BareMetalHost) error {
+	if metav1.GetControllerOf(machine) == nil {
+		log.Printf("Skipping deletion of machine %v: no owning controller", machine.Name)
+		return nil
+	}
+	if baremetalhost.Spec.ExternallyProvisioned {
+		log.Printf("Skipping deletion of machine %v: BMH %v is externally provisioned", machine.Name, baremetalhost.Name)
+		return nil
+	}
```
These checks should be outside this function.
Otherwise, we block remediation forever, even the reboot one, and we will constantly reach this.
If we can't reprpovision we should at least keep trying the powert-cycle flow.

We should do something like the isMaster() check. Perhaps do all the checks in a `canReprovision()` function that will check all conditions (non-master, has owning contoller and not externally provisioned)

ondrejmular replied:
AFAIK, power-cycle flow won't be affected as there is no change done until the machine is deleted.

But I like your idea creating a separate function for theses checks.

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -930,6 +1062,7 @@ The full remediation flow is:
 7) Remove poweredOffForRemediation annotation, MAO's machine unhealthy annotation and annotations/labels backup
 */
 func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta1.Machine, baremetalhost *bmh.BareMetalHost) error {
+	var mhc *machinev1beta1.MachineHealthCheck
```
nit: This line can be omitted and replace `mhc = ...` with `mhc := ...`

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -883,6 +909,112 @@ func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMe
 	return err
 }
 
+func isActionTimedOut(machine *machinev1beta1.Machine, annotation string, timeout time.Duration) bool {
```
would be good to add some documentation for this function

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -883,6 +909,112 @@ func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMe
+	return machine.Labels[machineRoleLabel] == machineRoleMaster
+}
+
+func (a *Actuator) getMhcByMachine(machine *machinev1beta1.Machine) *machinev1beta1.MachineHealthCheck {
```
same here for documentation

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -883,6 +909,112 @@ func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMe
+	return nil
+}
+
+func getTimeoutDurationFromAnnotation(mhc *machinev1beta1.MachineHealthCheck, annotation string, defaultDuration time.Duration) time.Duration {
```
documentation

n1r1 commented on pkg/cloud/baremetal/actuators/machine/actuator.go:
```diff
@@ -976,14 +1115,54 @@ func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta
+			mhc = a.getMhcByMachine(machine)
+			timeout := powerOnDefaultTimeout
+			if mhc != nil {
+				if duration, err := time.ParseDuration(mhc.Spec.NodeStartupTimeout); err == nil {
```
`mhc.Spec.NodeStartupTimeout` can be empty string, and this is a valid case.
In that case the log output might be a bit confusing.

ondrejmular replied:
ack

n1r1 reviewed (commented):
Looks good! 
I've mostly added some nits.

ondrejmular committed:
Improve error handling of machine remediation

Try to reprovision a machine (if possible) in case we are unable to
resolve issues just by rebooting it.
